Setup ocaml-freestanding as a cross-compiler

TheLortex opened the pull request:
## Changes

With this patch, ocaml-freestanding become a standalone ocamlfind toolchain. It can be used to build applications that aims to be linked with freestanding bindings. The compiler can be used by `dune` by using `(toolchain freestanding)` in the `context` stanza of `dune-workspace` files (https://dune.readthedocs.io/en/stable/dune-files.html#context) or with ocamlfind, by adding `-toolchain freestanding` in the command line. 

The cross-toolchain, nolibc and openlibm are installed in `<PREFIX>/freestanding-sysroot/`, acting as a sub-switch. It is discovered by ocamlfind and dune by installing a `freestanding.conf` file in `<PREFIX>/lib/findlib.conf.d/`. For backwards-compatibility, an `ocaml-freestanding.pc` file remains installed in `<PREFIX>/lib/pkgconfig/`. 

In the `test/` directory I included a minimal dune project aimed to be compiled with ocaml-freestanding. It's tested using Github actions for every solo5 bindings, in particular it runs the spt executable. FreeBSD is testing using cirrus ci.  

The changes depends on this Solo5 PR: https://github.com/Solo5/solo5/pull/488. This PR allows co-installability of solo5 targets that have the same set of C compilation flags.

## Advantages

We obtain a cross-compilation workflow, Mirage 4 using it to build unikernels (https://github.com/mirage/mirage/issues/1195). 
As host and unikernel compilers are decoupled, this means more customizability in the future as we could enable using different sets of compiler options without modifying the host switch. For now this would be done by having `ocaml-freestanding+<options>` variants of the package (like the compiler) but there is work on how to limit the combinatorial explosion: https://discuss.ocaml.org/t/experimental-new-layout-for-the-ocaml-variants-packages-in-opam-repository/6779

## Issues

Maintainability: the ocaml build system doesn't properly support cross-compilation so the Makefile is a bit hacky.

Build time: the initial build time will be longer, but the package need to be installed once for every solo5 target that support the same set of flags. 

Compiler options: this PR hardcodes the compiler options, but this could be added later.

## 

I hope everything makes sense, but I'm open to changes if something doesn't feel right! By the way I haven't updated documentation for now. 

PS: you can see the tests logs here: https://github.com/TheLortex/ocaml-freestanding/actions/runs/469376754

TheLortex commented:
Update: the history is a bit messy as I tried to add FreeBSD support. But I've been able to set up Cirrus CI and it's now green.

dinosaure commented:
Thanks for the cleaning, commits are really clear now :+1:.

TheLortex commented:
Thanks to the last fixes in https://github.com/Solo5/solo5/pull/488, this PR should be testable and reviewable. I've updated the PR description accordingly. 
@mato @hannesm @dinosaure I invite you to test it, and make sure it corresponds to what you have in mind when we discussed the changes. 

The `test/` directory is a good example on how OCaml code can be built with ocaml-freestanding using dune cross-compilation features. Use `MODE=<hvt|spt|...> dune build` to generate the application artifact in `_build/freestanding/main.out`.

mato commented on ocaml-freestanding.pc.in:
```diff
@@ -1,12 +1,10 @@
-Cflags: -I${includedir} -include _freestanding/overrides.h
-Libs: -L${libdir} -lasmrun -lnolibc -lopenlibm @@PKG_CONFIG_EXTRA_LIBS@@
-Requires: @@PKG_CONFIG_DEPS@@
+Cflags: -I${libdir} -I${sysroot}/include/nolibc
```
You are missing `-include _freestanding/overrides.h` here, see 239263b068e78f12123bfb526bb9b15d14f64a85 for why that exists.

mato commented on patches/configure.ac.patch:
```diff
@@ -0,0 +1,11 @@
+   [exeext=''])
+
+-otherlibraries="dynlink"
++otherlibraries=""
```
This is a bit unfortunate, since we now need `autoconf` and re-generating `configure` can be somewhat fragile. Can we eventually submit something to OCaml?

Alternatively -- this patch is likely to be fairly specific to the compiler version. If we see this becoming a problem, we could include version-specific patches for the _generated_ `configure` instead.

TheLortex replied:
I found another way of patching the build system that should be more robust across ocaml versions. The patches are simple enough so that we can simply edit the generated `configure` file. So I removed the dependency on `autoconf` and added `sed` transformations in the Makefile -- they perform simple replacements in `ocaml/configure`. 

The patches should be integrated in upstream OCaml at some point but I haven't took the time to look at the implications.

mato commented on Makefile:
```diff
@@ -35,77 +41,84 @@ ocaml/Makefile:
+		-disable-unix-lib\
+		-disable-instrumented-runtime
+	echo "ARCH=$(MAKECONF_OCAML_BUILD_ARCH)" >> ocaml/Makefile.config
+	echo '#define HAS_GETTIMEOFDAY' > ocaml/runtime/caml/s.h
```
Are you intentionally re-creating (`>` vs. `>>`) ocaml/runtime/caml/s.h here? If so, why?

TheLortex replied:
Yes. The point is that `runtime/caml/s.h` describes the _system_ configuration, and that configuration only depends on what's implemented in `nolibc`.
`./configure` generates `s.h` according to the host configuration, so it shouldn't be taken into account. 

Do you agree with that ?

hannesm replied:
My take is that since s.h depends on the target (unikernel with openlibm and nolibc), and not at all on the host system / architecture, we can safely ship our own `s.h` (in contrast to `m.h`). This means: we don't need to `echo` stuff in Makefile, but instead we should go back (e.g. to https://github.com/mirage/ocaml-freestanding/blob/ad125072c3dc2547b23c50f349cb83bf1bdb1134/config.in/s.h) and push an accurate `s.h` to ocaml-freestanding.

From reading an OCaml 4.12 s.h, we likely want to define at least HAS_C99_FLOAT_OPS and HAS_WORKING_FMA -- and need to figure out remaining defines (depending on the `ifdef` of the OCaml runtime code). There is certainly a maintainability concern, since upgrading OCaml may involve to upgrade s.h (while with `configure` and "editing" s.h (by appending via `echo ... >> s.h`) it looks less brittle).

hannesm replied:
Another thought: can't we execute `configure` with the right flags so that the headers from openlibm & nolibc are used (instead of the system headers)? If this is possible we would not need to manually edit / mess around with `s.h`.

mato replied:
> Another thought: can't we execute configure with the right flags so that the headers from openlibm & nolibc are used (instead of the system headers)? 

That should be possible, and it would indeed be the best way to get things to work and stay maintainable. @TheLortex ?

hannesm replied:
And indeed that is what `./configure CFLAGS=...` already does (by passing `CC=... $(OCAML_CFLAGS)`). But configure stumbles upon the issue:
```
configure:14095: checking for times
configure:14095: cc -ffreestanding -fstack-protector-strong -nostdlibinc -isystem /usr/home/hannes/.opam/4.11.1/lib/pkgconfig/../../include/solo5-bindings-xen/crt -I/usr/home/hannes/.opam/4.11.1/lib/pkgconfig/../../include/solo5-bindings-xen/solo5  -I/usr/home/hannes/devel/mirage/ocaml-freestanding/nolibc/include -include _freestanding/overrides.h -I/usr/home/hannes/devel/mirage/ocaml-freestanding/openlibm/include -I/usr/home/hannes/devel/mirage/ocaml-freestanding/openlibm/src -nostdlib -o conftest -g -O2 -ffreestanding -fstack-protector-strong -nostdlibinc -isystem /usr/home/hannes/.opam/4.11.1/lib/pkgconfig/../../include/solo5-bindings-xen/crt -I/usr/home/hannes/.opam/4.11.1/lib/pkgconfig/../../include/solo5-bindings-xen/solo5  -I/usr/home/hannes/devel/mirage/ocaml-freestanding/nolibc/include -include _freestanding/overrides.h -I/usr/home/hannes/devel/mirage/ocaml-freestanding/openlibm/include -I/usr/home/hannes/devel/mirage/ocaml-freestanding/openlibm/src  conftest.c -lm  >&5
ld: error: undefined symbol: times
>>> referenced by conftest.c:72
>>>               /tmp/conftest-1828cc.o:(main)
```

so, the implementation of (e.g.) times needs to be provided somehow (adding `sysdeps_solo5.o` to `LDFLAGS` -- which I couldn't work out (tried for 5 minutes)?).

mato replied:
> Eventually a "solo5-dummy" object just for the purpose of running configure successfully would be the way to go?

Sounds plausible. I'll take a look at this later today.

TheLortex replied:
Using a `solo5-dummy` object stubbing the solo5 API, I've been able to correctly set `configure` to produce a correct `s.h` !
So this is probably the most maintainable way. I'll update the PR

mato replied:
> Using a solo5-dummy object stubbing the solo5 API, I've been able to correctly set configure to produce a correct s.h !

Great. If it"s in any way cleaner, we could have a `sysdeps_stub.o` instead with stub implementations of the symbols defined there. However, I don't have any strong opinion about which is better, so take your pick.

mato commented on Makefile:
```diff
@@ -35,77 +41,84 @@ ocaml/Makefile:
+		CPPFLAGS="$(OC_CFLAGS)" \
+		LD="ld" \
+	  ./configure \
+		-target=$(shell gcc -dumpmachine) \
```
Why do we need to specify `-target` here? Also, this should definitely not be `gcc`; see other comments about `$CC`.

TheLortex replied:
Indeed it's not needed it's some remains from my experiments with the ocaml build system.

mato commented on Makefile:
```diff
@@ -1,22 +1,28 @@
+# CFLAGS used to build nolibc / openlibm / ocaml runtime
+LOCAL_CFLAGS=$(MAKECONF_CFLAGS) -I$(TOP)/nolibc/include -include _freestanding/overrides.h
+# CFLAGS used by the OCaml compiler to build C stubs
+GLOBAL_CFLAGS=$(MAKECONF_CFLAGS) -I$(MAKECONF_PREFIX)/freestanding-sysroot/include/nolibc/
```
This also needs `-include _freestanding/overrides.h`. Any C code compiled against ocaml-freestanding should use that (see other comment).

mato reviewed (commented):
This is looking good, and works as advertised, insofar as the link test in `test/` goes.

I've left a bunch of comments, hopefully not repeating myself too much. Please ask if anything is unclear.

Also, could you put together a PR to `mirage-solo5` to use the new freestanding compiler to build `libmirage-solo5_bindings.a`, i.e. revert the relevant parts of https://github.com/mirage/mirage-solo5/pull/72/commits/b871937382f84a6a9c7386b878ed996c4d770ec5? I'd try and do it myself, but you'll be quicker at the dune-fu.

mato commented:
General comment not directly related to the code review: Could you document somewhere what components (runtime headers, libraries, nolibc) are installed where under $PREFIX? Perhaps in the commit message for 22777f3, though a brief list would also be good to have in README.md.

Also, 89e1d05 has a typo "spl". Perhaps add that it is preparation for the cross-compiler work.

mato commented:
And one more thing, during build there are a bunch of warnings about "implicit declaration of function ..." while building parts of the compiler (bytecode interpreter?).

Do we use these parts of the compiler somehow? If not, why are we building them? If yes, these warnings should be addressed.

TheLortex commented on opam:
```diff
@@ -6,18 +6,28 @@ bug-reports: "https://github.com/mirage/ocaml-freestanding/issues/"
-substs: [
-  "flags/cflags.tmp"
-  "flags/libs.tmp"
+pin-depends: [
```
It is just for development purposes, they won't be needed once solo5 0.7.0 will be out. We can remove them before the PR is merged.

TheLortex commented on configure.sh:
```diff
@@ -44,17 +81,26 @@ case "${BUILD_ARCH}" in
         ;;
 esac
 
+case "$(cc -dumpmachine)" in
```
Fixed it.
Dune uses the compiler given by `ocamlc -config` so it would work as expected.

TheLortex commented on Makefile:
```diff
@@ -1,22 +1,28 @@
-.PHONY: all clean install uninstall distclean
+.PHONY: all clean install uninstall distclean ocaml
```
The ocaml build system is parallel but somewhat fragile. I added some sequentiality in the `ocaml` rule to allow parallel builds.

TheLortex commented:
Hey, thank you for your review ! I added individual commits addressing each point you mentioned. I'll rebase and squash them later, and address the remaining issues. I'll take care of integrating the changes in `mirage-solo5`. 

> Could you document somewhere what components (runtime headers, libraries, nolibc) are installed where under $PREFIX? 

The documentation will be updated accordingly. For reference: everything is installed in `$PREFIX/freestanding-sysroot/`. The runtime lives in `$PREFIX/freestanding-sysroot/lib/ocaml/` and nolibc lives in `$PREFIX/freestanding-sysroot/lib/nolibc/`. 

> And one more thing, during build there are a bunch of warnings about "implicit declaration of function ..." while building parts of the compiler (bytecode interpreter?).

The warnings comes from the build of `libcamlrun.a`, which may be useful to build bytecode executables and run them on a solo5 target. So I we want to support that we might need to implement the missing C functions or at least stub them out (`fputs`, `qsort`, `longjmp`, `strncmp`,  `putchar`).

mato commented:
> Hey, thank you for your review ! I added individual commits addressing each point you mentioned. I'll rebase and squash them later, and address the remaining issues. I'll take care of integrating the changes in `mirage-solo5`.

Great, thanks!

> The warnings comes from the build of `libcamlrun.a`, which may be useful to build bytecode executables and run them on a solo5 target. So I we want to support that we might need to implement the missing C functions or at least stub them out (`fputs`, `qsort`, `longjmp`, `strncmp`, `putchar`).

I'd not try to support bytecode now, we can look into that in a future release. Is it possible to disable building `libocamlrun.a` in the current version?

TheLortex commented:
@mato I've updated the PR. The most notable change is that the OCaml configure script link test is now working, so we don't need to override the generated `s.h`, except for `HAS_SOCKETS` because of a bug in the configure script. I also added some documentation, but feel free to ask for more precisions if needed.

mato commented:
@TheLortex This looks good to me, though pending a detailed review tomorrow. I've submitted a WIP mirage/mirage#1208 for the Mirage 3.x changes required.

dinosaure reviewed (approved):
I currently try to use `ocaml-freestanding` for the cross-compilation (for `aarch64`). Globally, this PR seems very good but I updated some part to really be able to use a `cc` cross-compiler (such as `aarch64-linux-gnu-gcc`). I think, but it seems orthogonal to this PR, we should let the user to define its own `$(CC)` (for `openlibm`, `nolibc` and `ocaml-freestanding`) and, by this way, unlock the ability to compile a "real" OCaml cross-compiler.

Anyway, it's a great job :tada:

mato commented:
On Thursday, 21.01.2021 at 07:37, Calascibetta Romain wrote:
> @dinosaure approved this pull request.
> 
> I currently try to use `ocaml-freestanding` for the cross-compilation (for `aarch64`). Globally, this PR seems very good but I updated some part to really be able to use a `cc` cross-compiler (such as `aarch64-linux-gnu-gcc`). I think, but it seems orthogonal to this PR, we should let the user to define its own `$(CC)` (for `openlibm`, `nolibc` and `ocaml-freestanding`) and, by this way, unlock the ability to compile a "real" OCaml cross-compiler.

That needs to be done from Solo5 up. There are some pieces already in place as part of solo5/solo5#488, such as `solo5-config --cc`, which will make this possible, and ocaml-freestanding already uses those, but not everything is in place right now. Patience :-)

TheLortex commented:
Superseded by https://github.com/mirage/ocaml-freestanding/pull/92
